### PR TITLE
[WIP]Refactor analyze and merge for crane agent

### PR DESCRIPTION
**Pull request by chenkaiyue**
Refactor crane agent

### CODE REVIEW

File: pkg/ensurance/analyzer/analyzer.go [hunk 1]
```diff
@@ -187,332 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
-			s.triggered[key] = 0
-			restored := utils.GetUint64FromMaps(key, s.restored)
-			restored++
-			s.restored[key] = restored
-			if restored >= uint64(utils.GetInt32withDefault(object.RestoreThreshold, known.DefaultRestoredThreshold)) {
-				dc.Restored = true
-			}
-		}
-	}
+	klog.V(4).Infof("DoAnalyze: key %s, threshold %v", key, threshold)
```
Comment (mfanjie) [inline pkg/ensurance/analyzer/analyzer.go:267]: suggest to remove the function name from log message `DoAnalyze `, as the log already show the code line number.

File: pkg/ensurance/analyzer/analyzer.go [hunk 2]
```diff
@@ -187,332 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
+
+	//step3: check is triggered action or restored, set the detection
+	s.actionTriggeredRestored(threshold, key, object, &dc, impacted)
 
 	return dc, nil
 }
 
-func (s *AnormalyAnalyzer) merge(stateMap map[string][]common.TimeSeries, avoidanceMaps map[string]*ensuranceapi.AvoidanceAction, dcs []ecache.DetectionCondition) executor.AvoidanceExecutor {
-	var now = time.Now()
+func (s *AnormalyAnalyzer) actionTriggeredRestored(threshold bool, key string, object ensuranceapi.ObjectiveEnsurance, dc *ecache.DetectionCondition, impacted []types.NamespacedName) {
```
Comment (mfanjie) [inline pkg/ensurance/analyzer/analyzer.go:275]: actionTriggeredRestored to be a more accurate function name?
how about computeActions?

File: pkg/ensurance/analyzer/analyzer.go [hunk 3]
```diff
@@ -187,332 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
-						enableSchedule = true
-						break
-					}
-				}
-			}
-		}
-	}
+	//step1 filter dry run detections
+	var dcsFiltered []ecache.DetectionCondition
+	dcsFiltered = s.filterDryRunDetections(dcs)
```
Comment (mfanjie) [inline pkg/ensurance/analyzer/analyzer.go:313]: I am wondering if we need filter dry run? as dry run is an action type IMO

File: pkg/ensurance/analyzer/analyzer.go [hunk 4]
```diff
@@ -187,332 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
+				influced := s.getImpacted(ts)
+				if len(influced) > 0 {
+					threshold = true
+					impacted = append(impacted, influced...)
 				}
 			} else {
-				// node metrics
-				klog.Infof("PodName or PodNamespace is empty")
 				threshold = true
+				return threshold, []types.NamespacedName{}
```
Comment (yan234280533) [inline pkg/ensurance/analyzer/analyzer.go:243]: not return here， return the end of the function ?

File: pkg/ensurance/analyzer/analyzer.go [hunk 5]
```diff
@@ -187,332 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
-	for _, t := range throttlePods {
-		if i := ae.ThrottleExecutor.ThrottleDownPods.Find(t.PodTypes); i == -1 {
-			ae.ThrottleExecutor.ThrottleDownPods = append(ae.ThrottleExecutor.ThrottleDownPods, t)
-		} else {
-			if t.CPUThrottle.MinCPURatio > ae.ThrottleExecutor.ThrottleDownPods[i].CPUThrottle.MinCPURatio {
-				ae.ThrottleExecutor.ThrottleDownPods[i].CPUThrottle.MinCPURatio = t.CPUThrottle.MinCPURatio
-			}
+		//step3 do Throttle merge FilterAndSortThrottlePods
+		if action.Spec.Throttle != nil {
+			throttlePods, throttleUpPods := s.GetThrottleActionPods(enableSchedule, dc, action, stateMap)
```
Comment (mfanjie) [inline pkg/ensurance/analyzer/analyzer.go:327]: possible to combine GetThrottleActionPods and GetEvictActionPods? as they are do they same thing, to find the targets.
and GetEvictPods and GetThrottlePods looks more senses

File: pkg/ensurance/executor/evict.go [hunk 6]
```diff
@@ -84,3 +84,17 @@ func (e *EvictExecutor) Avoid(ctx *ExecuteContext) error {
...
+
+func (e *EvictExecutor) Deduplicate(evictPods, NoEvictPods EvictPods) {
+	for _, ep := range evictPods {
+		if i := e.EvictPods.Find(ep.PodKey); i == -1 {
+			e.EvictPods = append(e.EvictPods, ep)
+		} else {
+			if ep.DeletionGracePeriodSeconds < e.EvictPods[i].DeletionGracePeriodSeconds {
+				e.EvictPods[i].DeletionGracePeriodSeconds = ep.DeletionGracePeriodSeconds
+			} else {
+				// nothing to do,replicated filter the evictPod
```
Comment (mfanjie) [inline pkg/ensurance/executor/evict.go:96]: if nothing to do, should you remove the else

File: pkg/ensurance/executor/throttle.go [hunk 7]
```diff
@@ -140,7 +141,7 @@ func (t *ThrottleExecutor) Avoid(ctx *ExecuteContext) error {
 			}
 
 			if limitCPU, ok := container.Resources.Limits[v1.ResourceCPU]; ok {
-				if float64(limitCPU.MilliValue())/1000.0*float64(throttlePod.CPUThrottle.MinCPURatio) > containerCPUQuotaNew {
+				if float64(limitCPU.MilliValue())/1000.0*float64(throttlePod.CPUThrottle.MinCPURatio)/100 > containerCPUQuotaNew {
```
Comment (mfanjie) [inline pkg/ensurance/executor/throttle.go:144]: can 1000.0 and 100 to be defined as const?

File: pkg/ensurance/analyzer/analyzer.go [hunk 8]
```diff
@@ -187,332 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
-			s.restored[key] = restored
-			if restored >= uint64(utils.GetInt32withDefault(object.RestoreThreshold, known.DefaultRestoredThreshold)) {
-				dc.Restored = true
-			}
-		}
-	}
+	klog.V(4).Infof("DoAnalyze: key %s, threshold %v", key, threshold)
+
+	//step3: check is triggered action or restored, set the detection
+	s.actionTriggeredRestored(threshold, key, object, &dc, impacted)
```
Comment (yan234280533) [inline pkg/ensurance/analyzer/analyzer.go:270]: getSeries,trigger, actionTriggeredRestored 

 the funtion name style must be consistent

File: pkg/ensurance/analyzer/analyzer.go [hunk 9]
```diff
@@ -187,332 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
+		restored++
+		s.restored[key] = restored
+		if restored >= uint64(utils.GetInt32withDefault(object.RestoreThreshold, known.DefaultRestoredThreshold)) {
+			dc.Restored = true
+		}
+	}
+}
 
-	//step1 filter dry run detections
+func (s *AnormalyAnalyzer) filterDryRunDetections(dcs []ecache.DetectionCondition) []ecache.DetectionCondition {
```
Comment (yan234280533) [inline pkg/ensurance/analyzer/analyzer.go:296]: filterDryRunDetections -> filterDryRun

File: pkg/ensurance/analyzer/analyzer.go (hunk 3 above)
Comment (yan234280533) [inline pkg/ensurance/analyzer/analyzer.go:313]: var dcsFiltered []ecache.DetectionCondition
	dcsFiltered = s.filterDryRunDetections(dcs)
->
dcsFiltered := s.filterDryRunDetections(dcs)

File: pkg/ensurance/analyzer/analyzer.go [hunk 10]
```diff
@@ -187,332 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
-	}
+	//step1 filter dry run detections
+	var dcsFiltered []ecache.DetectionCondition
+	dcsFiltered = s.filterDryRunDetections(dcs)
 
-	if enableSchedule {
-		ae.ScheduleExecutor.RestoreClassAndPriority = &executor.ClassAndPriority{PodQOSClass: v1.PodQOSBestEffort, PriorityClassValue: 0}
-	}
+	//step2 do DisableScheduled merge
+	enableSchedule := s.ScheduleMerge(dcsFiltered, avoidanceMaps, &ae)
```
Comment (yan234280533) [inline pkg/ensurance/analyzer/analyzer.go:316]: ScheduleMerge -> disableSchedulingMerge

File: pkg/ensurance/analyzer/analyzer.go [hunk 11]
```diff
@@ -187,332 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
-		if i := ae.ThrottleExecutor.ThrottleDownPods.Find(t.PodTypes); i == -1 {
-			ae.ThrottleExecutor.ThrottleDownPods = append(ae.ThrottleExecutor.ThrottleDownPods, t)
-		} else {
-			if t.CPUThrottle.MinCPURatio > ae.ThrottleExecutor.ThrottleDownPods[i].CPUThrottle.MinCPURatio {
-				ae.ThrottleExecutor.ThrottleDownPods[i].CPUThrottle.MinCPURatio = t.CPUThrottle.MinCPURatio
-			}
+		//step3 do Throttle merge FilterAndSortThrottlePods
+		if action.Spec.Throttle != nil {
+			throttlePods, throttleUpPods := s.GetThrottleActionPods(enableSchedule, dc, action, stateMap)
+			// combine the replicated pod
```
Comment (yan234280533) [inline pkg/ensurance/analyzer/analyzer.go:328]: here is throttle merge?

File: pkg/ensurance/analyzer/analyzer.go [hunk 12]
```diff
@@ -187,332 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
+			ae.ThrottleExecutor.Deduplicate(throttlePods.(executor.ThrottlePods), throttleUpPods.(executor.ThrottlePods))
+		}
 
-			if t.CPUThrottle.StepCPURatio > ae.ThrottleExecutor.ThrottleDownPods[i].CPUThrottle.StepCPURatio {
-				ae.ThrottleExecutor.ThrottleDownPods[i].CPUThrottle.StepCPURatio = t.CPUThrottle.StepCPURatio
-			}
+		//step4 do Evict merge FilterAndSortEvictPods
+		if action.Spec.Eviction != nil {
+			evictPods, _ := s.GetEvictActionPods(dc.Triggered, action)
+			// remove the replicated pod
```
Comment (yan234280533) [inline pkg/ensurance/analyzer/analyzer.go:335]: here is evict merge?

File: pkg/ensurance/analyzer/analyzer.go [hunk 13]
```diff
@@ -187,332 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
-	if !ok {
-		return dc, fmt.Errorf("metric %s not found", object.MetricRule.Name)
-	}
-	if strings.HasPrefix(object.MetricRule.Name, "container") {
-		//step1: get series from value
-
-		series := s.getTimeSeriesFromMap(state, object.MetricRule.Selector)
+func (s *AnormalyAnalyzer) getImpacted(ts common.TimeSeries) []types.NamespacedName {
+	//TODO: basicThrottleQosPriority be a input para to be a vara in AnormalyAnalyzer
+	var basicThrottleQosPriority = executor.ClassAndPriority{PodQOSClass: v1.PodQOSBestEffort, PriorityClassValue: 0}
```
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply pkg/ensurance/analyzer/analyzer.go:192]: I have moved the container's metric out, so there will don't have the baseline and this func has gone.

File: pkg/ensurance/analyzer/analyzer.go [hunk 14]
```diff
@@ -187,332 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
+	var impacted []types.NamespacedName
 
-		if len(series) == 0 {
-			return dc, fmt.Errorf("time series len is 0 for metric %s", object.MetricRule.Name)
-		}
+	pod, err := s.podLister.Pods(common.GetValueByName(ts.Labels, common.LabelNamePodNamespace)).Get(common.GetValueByName(ts.Labels, common.LabelNamePodName))
+	if err != nil {
+		klog.V(4).Infof("Warning: analyze: Pod %s/%s not found", common.GetValueByName(ts.Labels, common.LabelNamePodNamespace), common.GetValueByName(ts.Labels, common.LabelNamePodName))
```
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply pkg/ensurance/analyzer/analyzer.go:197]: Done

**Comment by chenkaiyue** [comment]
To be accomplished:
After an action is triggered, we will **don't have the baseline** to filter pods, we will select every pod as an alternative;
Use more dimensional metrics to sort pods;
In executor, we will calculate the diff to the water level, and select some pods from the sorted pods according to the diff and end the process.

File: pkg/ensurance/executor/evict.go [hunk 15]
```diff
@@ -84,3 +84,17 @@ func (e *EvictExecutor) Avoid(ctx *ExecuteContext) error {
 func (e *EvictExecutor) Restore(ctx *ExecuteContext) error {
 	return nil
 }
+
+func (e *EvictExecutor) Deduplicate(evictPods, NoEvictPods EvictPods) {
```
[the remark this answers is not in the document]
Comment (mfanjie) [reply pkg/ensurance/executor/evict.go:88]: deduplicate is a correct word

File: pkg/ensurance/analyzer/analyzer.go [hunk 16]
```diff
@@ -187,332 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
 				}
 			} else {
-				// node metrics
-				klog.Infof("PodName or PodNamespace is empty")
 				threshold = true
+				return threshold, []types.NamespacedName{}
 			}
 		}
+	}
+	return threshold, impacted
```
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply pkg/ensurance/analyzer/analyzer.go:247]: Done

File: pkg/ensurance/analyzer/analyzer.go (hunk 1 above)
Comment (chenkaiyue) [reply pkg/ensurance/analyzer/analyzer.go:267]: Done

File: pkg/ensurance/analyzer/analyzer.go [hunk 17]
```diff
@@ -187,332 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
-			s.triggered[key] = 0
-			restored := utils.GetUint64FromMaps(key, s.restored)
-			restored++
-			s.restored[key] = restored
-			if restored >= uint64(utils.GetInt32withDefault(object.RestoreThreshold, known.DefaultRestoredThreshold)) {
-				dc.Restored = true
-			}
-		}
+func (s *AnormalyAnalyzer) analyze(key string, object ensuranceapi.ObjectiveEnsurance, stateMap map[string][]common.TimeSeries) (ecache.DetectionCondition, error) {
+	var dc = ecache.DetectionCondition{Strategy: object.Strategy, ObjectiveEnsuranceName: object.Name, ActionName: object.AvoidanceActionName}
```
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply pkg/ensurance/analyzer/analyzer.go:251]: Rename it to ActionContext

File: pkg/ensurance/analyzer/analyzer.go (hunk 2 above)
Comment (chenkaiyue) [reply pkg/ensurance/analyzer/analyzer.go:275]: OK, rename to computeActionContext.

File: pkg/ensurance/executor/evict.go (hunk 6 above)
Comment (chenkaiyue) [reply pkg/ensurance/executor/evict.go:96]: Done

File: pkg/ensurance/executor/throttle.go (hunk 7 above)
Comment (chenkaiyue) [reply pkg/ensurance/executor/throttle.go:144]: Done

File: pkg/ensurance/executor/evict.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply pkg/ensurance/executor/evict.go:88]: OK, rename to CombineDuplicate.

File: pkg/ensurance/collector/collector.go [hunk 18]
```diff
@@ -145,16 +144,15 @@ func (s *StateCollector) UpdateCollectors() {
...
-
-		// cadvisor need to stop manager before
-		if value, exists := s.collectors.Load(types.CadvisorCollectorType); !exists {
-			s.collectors.Delete(types.CadvisorCollectorType)
-			c := value.(Collector)
-			if err = c.Stop(); err != nil {
-				klog.Errorf("Failed to stop the cadvisor manager.")
+		CollectorsNeedToStop := []types.CollectType{types.NodeLocalCollectorType, types.CadvisorCollectorType}
+
+		for _, collector := range CollectorsNeedToStop{
```
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply pkg/ensurance/collector/collector.go:149]: Done

File: pkg/ensurance/analyzer/analyzer.go [hunk 19]
```diff
@@ -602,25 +416,122 @@ func (s *AnormalyAnalyzer) actionTriggered(dc ecache.DetectionCondition) bool {
...
+	if enableSchedule {
+		ae.ScheduleExecutor.RestoreClassAndPriority = &executor.ClassAndPriority{PodQOSClass: v1.PodQOSBestEffort, PriorityClassValue: 0}
+	} else {
+		ae.ScheduleExecutor.DisableClassAndPriority = &executor.ClassAndPriority{PodQOSClass: v1.PodQOSBestEffort, PriorityClassValue: 0}
+	}
+	s.lastTriggeredTime = now
+	return
+}
+
+func throttlePodConstruct(qosPriority executor.ClassAndPriority, pod *v1.Pod, stateMap map[string][]common.TimeSeries, action *ensuranceapi.AvoidanceAction) executor.ThrottlePod{
```
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply pkg/ensurance/analyzer/analyzer.go:523]: Done

File: pkg/ensurance/analyzer/analyzer.go [hunk 20]
```diff
@@ -187,332 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
-	var dc = ecache.DetectionCondition{Strategy: object.Strategy, ObjectiveEnsuranceName: object.Name, ActionName: object.AvoidanceActionName}
-	state, ok := stateMap[object.MetricRule.Name]
-	if !ok {
-		return dc, fmt.Errorf("metric %s not found", object.MetricRule.Name)
-	}
-	if strings.HasPrefix(object.MetricRule.Name, "container") {
-		//step1: get series from value
-
-		series := s.getTimeSeriesFromMap(state, object.MetricRule.Selector)
+func (s *AnormalyAnalyzer) getImpacted(ts common.TimeSeries) []types.NamespacedName {
```
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply pkg/ensurance/analyzer/analyzer.go:190]: OK, I have moved the container's metric out, so this func has gone.

File: pkg/ensurance/analyzer/analyzer.go (hunk 4 above)
Comment (chenkaiyue) [reply pkg/ensurance/analyzer/analyzer.go:243]: Done

File: pkg/ensurance/executor/throttle.go [hunk 21]
```diff
@@ -261,3 +262,56 @@ func (t *ThrottleExecutor) Restore(ctx *ExecuteContext) error {
 
 	return nil
 }
+
+func (e *ThrottleExecutor) Deduplicate(throttlePods, throttleUpPods ThrottlePods) {
```
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply pkg/ensurance/executor/throttle.go:266]: Made it a func in analyzer

File: pkg/ensurance/analyzer/analyzer.go (hunk 9 above)
Comment (chenkaiyue) [reply pkg/ensurance/analyzer/analyzer.go:296]: Done

File: pkg/ensurance/analyzer/analyzer.go (hunk 3 above)
Comment (chenkaiyue) [reply pkg/ensurance/analyzer/analyzer.go:313]: Done

File: pkg/ensurance/analyzer/analyzer.go (hunk 10 above)
Comment (chenkaiyue) [reply pkg/ensurance/analyzer/analyzer.go:316]: Done

File: pkg/ensurance/analyzer/analyzer.go (hunk 11 above)
Comment (chenkaiyue) [reply pkg/ensurance/analyzer/analyzer.go:328]: fix the comment to
"get and deduplicate throttlePods, throttleUpPods"

File: pkg/ensurance/analyzer/analyzer.go (hunk 12 above)
Comment (chenkaiyue) [reply pkg/ensurance/analyzer/analyzer.go:335]: fix the comment to
"get and deduplicate evictPods"

File: pkg/ensurance/analyzer/analyzer.go [hunk 22]
```diff
@@ -187,332 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
-			klog.V(6).Infof("Anormaly detection result %v, Name: %s, Value: %.2f, %s/%s", triggered,
-				object.MetricRule.Name,
-				ts.Samples[0].Value,
-				common.GetValueByName(ts.Labels, common.LabelNamePodNamespace),
-				common.GetValueByName(ts.Labels, common.LabelNamePodName))
+func (s *AnormalyAnalyzer) trigger(series []common.TimeSeries, object ensuranceapi.ObjectiveEnsurance) (bool, []types.NamespacedName) {
+	var impacted []types.NamespacedName
+	var triggered, threshold bool
+	for _, ts := range series {
+		triggered = s.evaluator.EvalWithMetric(object.MetricRule.Name, float64(object.MetricRule.Value.Value()), ts.Samples[0].Value)
```
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply pkg/ensurance/analyzer/analyzer.go:206]: To be accomplished:
1. After an action is triggered, we will **don't have the baseline** to filter pods, we will select every pod as an alternative;
2. Use **more dimensional metrics to sort** pods;
3. In executor, we will calculate the diff to the water level, and select some pods from the sorted pods **according to the diff and end the process**.

File: pkg/ensurance/analyzer/analyzer.go [hunk 23]
```diff
@@ -187,383 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
-				ts.Samples[0].Value,
-				common.GetValueByName(ts.Labels, common.LabelNamePodNamespace),
-				common.GetValueByName(ts.Labels, common.LabelNamePodName))
+	//step2: use opa to check if triggered and get impacted pods for container MetricRule
+	threshold := s.trigger(series, object)
 
-			if !triggered {
-				continue
-			}
+	klog.V(4).Infof("key %s, threshold %v", key, threshold)
```
Comment (mfanjie) [inline pkg/ensurance/analyzer/analyzer.go:233]: plz refine this log, log should be a sentence

File: pkg/ensurance/executor/throttle.go [hunk 24]
```diff
@@ -8,12 +8,15 @@ import (
...
 	"k8s.io/klog/v2"
 
+	"github.com/gocrane/crane/pkg/common"
 	cruntime "github.com/gocrane/crane/pkg/ensurance/runtime"
 	"github.com/gocrane/crane/pkg/utils"
 )
 
 const (
-	MAX_UP_QUOTA = 60 * 1000 // 60CU
+	MAX_UP_QUOTA          = 60 * 1000 // 60CU
```
Comment (mfanjie) [inline pkg/ensurance/executor/throttle.go:17]: suggest to use Camel style to define const as well.

**Comment by chenkaiyue** [comment]
image
image

File: pkg/ensurance/analyzer/analyzer.go [hunk 25]
```diff
@@ -187,383 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
+	if err != nil {
+		return ac, err
+	}
 
-			klog.V(6).Infof("Anormaly detection result %v, Name: %s, Value: %.2f, %s/%s", triggered,
-				object.MetricRule.Name,
-				ts.Samples[0].Value,
-				common.GetValueByName(ts.Labels, common.LabelNamePodNamespace),
-				common.GetValueByName(ts.Labels, common.LabelNamePodName))
+	//step2: use opa to check if triggered and get impacted pods for container MetricRule
```
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply pkg/ensurance/analyzer/analyzer.go:230]: Done

File: pkg/ensurance/analyzer/analyzer.go [hunk 26]
```diff
@@ -187,383 +187,161 @@ func (s *AnormalyAnalyzer) Analyze(state map[string][]common.TimeSeries) {
...
+	return ac, nil
+}
 
-		//step1: get metric value
-		value, err := s.getMetricFromMap(object.MetricRule.Name, state, object.MetricRule.Selector)
-		if err != nil {
-			return dc, err
+func (s *AnormalyAnalyzer) computeActionContext(threshold bool, key string, object ensuranceapi.ObjectiveEnsurance, ac *ecache.ActionContext) {
+	if threshold {
+		s.restored[key] = 0
```
[the remark this answers is not in the document]
Comment (chenkaiyue) [reply pkg/ensurance/analyzer/analyzer.go:243]: Yeah, this is not thread safe. Not only for restored, triggered and actionEventStatus are map two. I will make another PR to fix this.

**Comment by chenkaiyue** [comment]
For action throttle:
we use two pod:
image
image
and start a stress-ng process in middle,
action is
image

After a while, the middle pod cpu quota is
image
and the cpu usage on node is :
wecom-temp-5100d8d7a187a2a0165ab5b5413229d5
caculate cpuacct.usage changes in a period of time:
wecom-temp-67252523d752101d946d83af71d0675c

**Comment by mfanjie** [comment]
conditionType: analyzed-pressure should follow same style with existing ones, like `PIDPressure`.
and It's better if we can show what kind of pressure is identified instead of analyzed-pressure.

**Pull request merged**